File: Java/Route-Between-Two-Nodes-in-Graph.java
/**
 * Definition for Directed graph.
 * class DirectedGraphNode {
 *     int label;
 *     ArrayList<DirectedGraphNode> neighbors;
 *     DirectedGraphNode(int x) {
 *         label = x;
 *         neighbors = new ArrayList<DirectedGraphNode>();
 *     }
 * };
 */
public class Solution {
   /**
     * @param graph: A list of Directed graph node
     * @param s: the starting Directed graph node
     * @param t: the terminal Directed graph node
     * @return: a boolean value
     */
   public boolean hasRoute(ArrayList<DirectedGraphNode> graph,
                            DirectedGraphNode s, DirectedGraphNode t) {
        // write your code here
        Queue<DirectedGraphNode> q = new LinkedList<DirectedGraphNode>();
        HashSet<DirectedGraphNode> hash = new HashSet<DirectedGraphNode>();
        q.offer(s);
        hash.add(s);
        while (!q.isEmpty()) {
            DirectedGraphNode cur = q.poll();
            if (cur == t) return true;
            for (int i = 0; i < cur.neighbors.size(); i ++) {
                DirectedGraphNode temp = cur.neighbors.get(i);
                //if (temp == t) return true;
                if (!hash.contains(temp)) {
                    hash.add(temp);
                    q.offer(temp);
                }
            }
        }
        return false;
    }
}

File: Python/Route-Between-Two-Nodes-in-Graph.py
# Definition for a Directed graph node
# class DirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []

class Solution:
    """
    @param graph: A list of Directed graph node
    @param s: the starting Directed graph node
    @param t: the terminal Directed graph node
    @return: a boolean value
    """
    def hasRoute(self, graph, s, t):
        # write your code here
        q = []
        hash = {}
        hash[s] = 1
        q.append(s)
        while len(q) > 0:
            temp = q.pop(0)
            if temp is t:
                return True
            for item in temp.neighbors:
                if item not in hash:
                    hash[item] = 1
                    q.append(item)
        return False